extensions/ql-vscode/src/run-remote-query.ts

```diff
@@ -413,3 +422,116 @@ export async function attemptRerun(
+ */
+function getArtifactIDfromName(artifactName: string, artifacts: Array<{ id: number, name: string }>): number | undefined {
+  const artifact = artifacts.find(a => a.name.localeCompare(artifactName) === 0);
+  return artifact ? artifact.id : undefined;
+}
+
+/**
+ * Downloads an artifact from a workflow run.
+ * @param credentials Credentials for authenticating to the GitHub API.
+ * @param owner
+ * @param repo
+ * @param artifactId The ID of the artifact to download.
+ * @returns The path to the enclosing directory of the unzipped artifact.
+ */
+async function downloadArtifact(
+  credentials: Credentials,
+  owner: string,
+  repo: string,
+  artifactId: number
+): Promise<string> {
+  const octokit = await credentials.getOctokit();
+  const response = await octokit.rest.actions.downloadArtifact({
+    owner,
+    repo,
+    artifact_id: artifactId,
+    archive_format: 'zip',
+  });
+  const artifactPath = path.join(tmpDir.name, `${artifactId}`);
+  void logger.log(`Downloading artifact to ${artifactPath}.zip`);
+  await fs.writeFile(
+    `${artifactPath}.zip`,
+    Buffer.from(response.data as ArrayBuffer)
+  );
+
+  void logger.log(`Extracting artifact to ${artifactPath}`);
+  await (
+    await unzipper.Open.file(`${artifactPath}.zip`)
+  ).extract({ path: artifactPath });
+  return artifactPath;
+}
+
+interface ResultIndexItem {
+  nwo: string;
+  id: string;
+  results_count: number;
+  bqrs_file_size: number;
+  sarif_file_size?: number;
+}
+
+/**
+ * Gets the result index file for a given remote queries run.
+ * @param credentials Credentials for authenticating to the GitHub API.
+ * @param owner
+ * @param repo
+ * @param workflowRunId The ID of the workflow run to get the result index for.
+ * @returns An object containing the result index.
+ */
+export async function getResultIndex(
+  credentials: Credentials,
+  owner: string,
+  repo: string,
+  workflowRunId: number
+): Promise<ResultIndexItem[]> {
+  const artifactList = await listWorkflowRunArtifacts(credentials, owner, repo, workflowRunId);
+  const artifactId = getArtifactIDfromName('result-index', artifactList);
+  if (!artifactId) {
+    void showAndLogWarningMessage(
+      `Could not find a result index for the [specified workflow](https://github.com/${owner}/${repo}/actions/runs/${workflowRunId}).
+      Please check whether the workflow run has successfully completed.`
+    );
+    return [];
+  }
+  const artifactPath = await downloadArtifact(credentials, owner, repo, artifactId);
+  const indexFilePath = path.join(artifactPath, 'index.json');
+  if (!(await fs.pathExists(indexFilePath))) {
+    void showAndLogWarningMessage('Could not find an `index.json` file in the result artifact.');
+    return [];
+  }
+  const resultIndex = await fs.readFile(path.join(artifactPath, 'index.json'), 'utf8');
+
+  // Logging the result for testing purposes. This is temporary.
+  void logger.log(`Result index: ${resultIndex}`);
+  return JSON.parse(resultIndex);
```

Comment:
Can you wrap this in a try-catch? Rethrow the error with a more understandable user readable message? Maybe something like:

> "Invalid result-index file: " + {add the original message here}